virtualmachine/vsphere/util.go

```diff
@@ -108,6 +109,71 @@ var readAll = func(r io.Reader) ([]byte, error) {
 	return ioutil.ReadAll(r)
 }
 
+var extractOva = func(basePath string, body io.Reader) (string, error) {
+        tarBallReader := tar.NewReader(body)
+        var ovfFileName string
+
+        for {
+                header, err := tarBallReader.Next()
+                if err != nil {
+                        if err == io.EOF {
+                                break
+                        }
+                        return "", err
+                }
+                filename := header.Name
+                if filepath.Ext(filename) == ".ovf" {
+                        ovfFileName = filename
+                }
+                err = func() error {
+                        writer, err := os.Create(filepath.Join(basePath, filename))
+                        defer writer.Close()
+                        if err != nil {
+                                return err
+                        }
+                        _, err = io.Copy(writer, tarBallReader)
+                        if err != nil {
+                                return err
+                        }
+                        return nil
+		}()
+                if err != nil {
+                        return "", err
+                }
+        }
+        if ovfFileName == "" {
+                return "", errors.New("no ovf file found in the archive")
+        }
+        return filepath.Join(basePath, ovfFileName), nil
+}
+
+var downloadOva = func(basePath, url string) (string, error) {
+	var ovaReader io.Reader
+	if strings.HasPrefix(url, "http://") || strings.HasPrefix(url, "https://") {
+		resp, err := http.Get(url)
+		if err != nil {
+			return "", err
+		}
+		ovaReader = resp.Body
+		if resp.StatusCode != http.StatusOK {
+			return "", errors.New(fmt.Sprintf("can't download ova file from url: %s, status: %d", url, resp.StatusCode))
```

Comment:
I think fmt.Errorf("...") is sufficient to achieve this behavior. Please confirm.